File: content/snippets/python/design_patterns/object_pool.py
class PooledObject:
    def __init__(self):
        self.value = None

class ObjectPool:
    def __init__(self):
        self._pool = []
    def acquire(self):
        return self._pool.pop() if self._pool else PooledObject()
    def release(self, obj):
        self._pool.append(obj)

if __name__ == '__main__':
    p = ObjectPool()
    o = p.acquire()
    o.value = 1
    print(o.value)
    p.release(o)

File: content/snippets/typescript/design_patterns/object_pool.ts
class PooledObject { value: number | null = null }
class ObjectPool<T extends PooledObject> {
  private pool: T[] = []
  acquire(factory: () => T) { return this.pool.pop() ?? factory() }
  release(obj: T) { this.pool.push(obj) }
}

const pool = new ObjectPool<PooledObject>()
const o = pool.acquire(() => new PooledObject())
o.value = 7
console.log(o.value)
pool.release(o)
export {}
